tide-slog/src/per_request_logger.rs

```diff
@@ -0,0 +1,77 @@
+use futures::future::BoxFuture;
+use tide_core::{
+    middleware::{Middleware, Next},
+    Context, Response,
+};
+
+/// Middleware that injects a per-request [`slog::Logger`] onto the request [`Context`].
+pub struct PerRequestLogger<State> {
+    root_logger: slog::Logger,
+    setup:
+        Box<dyn (Fn(&slog::Logger, &mut Context<State>) -> slog::Logger) + Send + Sync + 'static>,
+}
+
+impl<State> PerRequestLogger<State> {
+    /// Initialize this middleware with a root logger to base the per-request loggers on and a
+    /// function to initialize those loggers with extra data.
+    ///
+    /// # Examples
+    ///
+    /// ## Adding a base16 encoded per-request UUID onto the logger context
+    ///
+    /// ```
+    /// use slog::o;
+    ///
+    /// let mut app = tide::App::new();
+    ///
+    /// let request_id = || uuid::Uuid::new_v4().to_simple().to_string();
+    ///
+    /// let root_logger = slog::Logger::root(slog::Discard, o!());
+    /// app.middleware(tide_slog::PerRequestLogger::new(root_logger, move |logger, _cx| logger.new(o! {
+    ///     "request" => request_id(),
+    /// })));
+    /// ```
+    ///
+    /// ## Taking an externally provided request id from headers for the logger context
+    ///
+    /// ```
+    /// use slog::o;
+    ///
+    /// let mut app = tide::App::new();
+    ///
+    /// let root_logger = slog::Logger::root(slog::Discard, o!());
+    /// app.middleware(tide_slog::PerRequestLogger::new(root_logger, move |logger, cx| logger.new(o! {
+    ///     "request" => cx.headers().get("Request-Id").unwrap().to_str().unwrap().to_owned(),
+    /// })));
+    /// ```
+    pub fn new(
+        root_logger: slog::Logger,
+        setup: impl (Fn(&slog::Logger, &mut Context<State>) -> slog::Logger) + Send + Sync + 'static,
+    ) -> Self {
+        Self {
+            root_logger,
+            setup: Box::new(setup),
```

Comment:
After going through all the changes, I see how the setup function can be used in the case where no additional per request setup needs to be performed. I still feel that it can be further simplified, where a default setup function is provided by the middleware, so that users who do not require the additional setup won't have to remember to make sure to pass `move |root_logger, _cx| root_logger.clone()` each time.

Reply:
I've replaced `new` with separate

```rust
    pub fn with_setup(
        setup: impl (Fn(&mut Context<State>) -> slog::Logger) + Send + Sync + 'static,
    )
    pub fn with_logger(logger: slog::Logger) -> Self
```

I realised that taking in the `root_logger` was actually useless as well since that can just be part of the `setup` closure environment.

Reply:
Yeah, this change makes it a lot cleaner. It looks good to merge to me.